### ISSUE TITLE
adjust mean annual precip processing to be more memory efficent

### ISSUE THREAD

**Issue by NewGraphEnvironment** [smnorris/bcfishpass#108]
having memory issues when trying to calculate mean annual precipitation for the whole province.  Managed to get `map` for most the watersheds for the province and now looking at the weighted mean which seems even more relevant (another nice use of `FWA_Upstream` btw).  For some reason my `ext4.vhdx` (image of my wsl2 drive) balloons when I run these large moves (my `bcfishpass.mean_annual_precip` table has over 2 million rows) and it fails when I try to do too much at once (adding over 150GB to the image file! - ha).  I need to break up the province and optimize the virtual drive image in between chunks (fun).  I wonder if the balloon effect might not be as pronounced if I run one watershed group at a time....

Can you think of a way to iterate through watershed groups rather than try to run a bunch at once?  The command I am running right now with 'BULK' as an example watershed is below.  I would like to pass the `WHERE a.watershed_group_code IN ....` line a list to be run in sequence.  Is that something you are familiar with when running raw psql commands? :

-- calculate the area weighted mean of precip contributing to a stream
`WITH areas AS
(SELECT
  a.id,
  b.map,
  b.area
FROM bcfishpass.mean_annual_precip a
INNER JOIN bcfishpass.mean_annual_precip b
ON FWA_Upstream(a.wscode_ltree, a.localcode_ltree, b.wscode_ltree, b.localcode_ltree)
WHERE a.watershed_group_code IN ('BULK')
),

totals AS
(
  SELECT
  a.id,
  SUM(area) as area
  FROM areas a
  GROUP BY a.id
),

weighting AS
(
SELECT
 a.id,
 a.area,
 b.area as area_total,
 a.area / b.area as pct,
 a.map,
 (a.area / b.area) * a.map as weighted_map
FROM areas a
INNER JOIN totals b ON a.id = b.id
),

weighted_mean AS
(
SELECT
  id,
  round(sum(weighted_map)) as map_upstream
FROM weighting
GROUP BY id
)

UPDATE bcfishpass.mean_annual_precip s
SET map_upstream = w.map_upstream
FROM weighted_mean w
WHERE s.id = w.id;`

**Comment by smnorris** [comment]
Yes, that wouldn't work very well provincially. 
Easiest fix is to make watershed group a variable and iterate through them in `mean_annual_precip.sh`.

**Comment by NewGraphEnvironment** [comment]
can you feed a `psql` command a variable or do you need to wrap it with `python` or `R`?

**Comment by smnorris** [comment]
Yup. Just looking around for places where I've already done it. `mean_annual_precip.sql` will have to be broken into a few pieces too.

**Comment by NewGraphEnvironment** [comment]
ya, i'm just running that one command as above with a modified version of `sql/mean_annual_precip.sql` 
`psql -f sql/mean_annual_precip2_erase.sql`

**Comment by NewGraphEnvironment** [comment]
I ran the Skeena and it only took a few minutes but it did tack like 15GB to my `wsl` image.  So annoying...

**Comment by smnorris** [comment]
Shouldn't be as much of an issue if running watershed groups individually. But for the mouth of the Skeena is it is a big calculation.

**Comment by smnorris** [comment]
Here is an example of how to do it: https://github.com/smnorris/fwapg/blob/main/extras/assessment_watersheds_lookups.sh#L13

If you have time to do it right now, go for it. I can probably get it done today too.

**Comment by NewGraphEnvironment** [comment]
Cool.  Basic lack of knowlege here... Can I define the variable in the line before as below?

`wsg = 'ADMS', 'ALBN', 'ATLL', 'BABL', 'BABR', 'BBAR', 'BEAV', 'BELA', 'BIGC', 'BLAR', 'BLUR', 'BONP', 'BOWR', 'BRID', 'BRKS', 'BULK', 'BULL', 'CAMB', 'CANO', 'CARP', 'CARR', 'CHES', 'CHIL', 'CHWK', 'CLAY', 'CLRH', 'CLWR', 'COLR', 'COTR', 'COWN', 'CRKD', 'DEAD', 'DEAL', 'DOGC', 'DRIR', 'DUNC', 'DUNE', 'ELKR', 'EUCH', 'EUCL', 'FINA', 'FINL', 'FIRE', 'FONT', 'FOXR', 'FRAN', 'FRCN', 'GATA', 'GLAR', 'GOLD', 'GRAI', 'GUIC', 'HARR', 'HERR', 'HOLB', 'HOMA', 'HORS', 'INGR', 'INKR', 'ISKR', 'JERV', 'KAKC', 'KCHL', 'KETL', 'KHOR', 'KHTZ', 'KINR', 'KISK', 'KISP', 'KITR', 'KLIN', 'KLUM', 'KNIG', 'KOTL', 'KOTR', 'KSHR', 'KUMR', 'LARL', 'LBIR', 'LBTN', 'LCHL', 'LCHR', 'LEUT', 'LFRA', 'LFRT', 'LHAF', 'LIAR', 'LILL', 'LISR', 'LKEL', 'LMUS', 'LNAR', 'LNIC', 'LNRS', 'LNTH', 'LOMI', 'LPCE', 'LPET', 'LRDO', 'LSAL', 'LSIK', 'LSKE', 'LTRE', 'MAHD', 'MCGR', 'MDEA', 'MESC', 'MESI', 'MFRA', 'MFRT', 'MIDR', 'MILL', 'MMUS', 'MORI', 'MORK', 'MORR', 'MPRO', 'MSKE', 'MSTR', 'MURR', 'MUSK', 'NAHR', 'NAKR', 'NARC', 'NASC', 'NASR', 'NATR', 'NAZR', 'NECL', 'NECR', 'NEVI', 'NICL', 'NIEL', 'NIMP', 'OKAN', 'OSPK', 'OWIK', 'PARA', 'PARK', 'PARS', 'PCEA', 'PINE', 'QUES', 'REVL', 'SAHT', 'SAJR', 'SALM', 'SALR', 'SANJ', 'SETN', 'SHER', 'SHUL', 'SIML', 'SLOC', 'SMAR', 'SMOK', 'SQAM', 'STHM', 'STUL', 'STUR', 'SUST', 'TABR', 'TAHS', 'TAKL', 'TAYR', 'THOM', 'TOAD', 'TOBA', 'TSAY', 'TSEA', 'TSIT', 'TURN', 'UARL', 'UBIR', 'UBTN', 'UCHR', 'UDEN', 'UEUT', 'UFRA', 'UHAF', 'UISR', 'UKEC', 'UMUS', 'UNAR', 'UNRS', 'UNTH', 'UNUR', 'UOMI', 'UPCE', 'USHU', 'USIK', 'USKE', 'UTRE', 'VICT', 'WILL', 'WORC', 'ZYMO'`

**Comment by NewGraphEnvironment** [comment]
like this perhaps?

-- Get mean annual precipitation for each stream segment,
-- both at the stream and the average for the entire contributing area

wsg = c('ADMS', 'ALBN', 'ATLL', 'BABL', 'BABR', 'BBAR', 'BEAV', 'BELA', 'BIGC', 'BLAR', 'BLUR', 'BONP', 'BOWR',
'BRID', 'BRKS', 'BULK', 'BULL', 'CAMB', 'CANO', 'CARP', 'CARR', 'CHES', 'CHIL', 'CHWK', 'CLAY', 'CLRH', 'CLWR',
'COLR', 'COTR', 'COWN', 'CRKD', 'DEAD', 'DEAL', 'DOGC', 'DRIR', 'DUNC', 'DUNE', 'ELKR', 'EUCH', 'EUCL', 'FINA',
'FINL', 'FIRE', 'FONT', 'FOXR', 'FRAN', 'FRCN', 'GATA', 'GLAR', 'GOLD', 'GRAI', 'GUIC', 'HARR', 'HERR', 'HOLB',
'HOMA', 'HORS', 'INGR', 'INKR', 'ISKR', 'JERV', 'KAKC', 'KCHL', 'KETL', 'KHOR', 'KHTZ', 'KINR', 'KISK', 'KISP',
'KITR', 'KLIN', 'KLUM', 'KNIG', 'KOTL', 'KOTR', 'KSHR', 'KUMR', 'LARL', 'LBIR', 'LBTN', 'LCHL', 'LCHR', 'LEUT',
'LFRA', 'LFRT', 'LHAF', 'LIAR', 'LILL', 'LISR', 'LKEL', 'LMUS', 'LNAR', 'LNIC', 'LNRS', 'LNTH', 'LOMI', 'LPCE',
'LPET', 'LRDO', 'LSAL', 'LSIK', 'LSKE', 'LTRE', 'MAHD', 'MCGR', 'MDEA', 'MESC', 'MESI', 'MFRA', 'MFRT', 'MIDR',
'MILL', 'MMUS', 'MORI', 'MORK', 'MORR', 'MPRO', 'MSKE', 'MSTR', 'MURR', 'MUSK', 'NAHR', 'NAKR', 'NARC', 'NASC',
'NASR', 'NATR', 'NAZR', 'NECL', 'NECR', 'NEVI', 'NICL', 'NIEL', 'NIMP', 'OKAN', 'OSPK', 'OWIK', 'PARA', 'PARK',
'PARS', 'PCEA', 'PINE', 'QUES', 'REVL', 'SAHT', 'SAJR', 'SALM', 'SALR', 'SANJ', 'SETN', 'SHER', 'SHUL', 'SIML',
'SLOC', 'SMAR', 'SMOK', 'SQAM', 'STHM', 'STUL', 'STUR', 'SUST', 'TABR', 'TAHS', 'TAKL', 'TAYR', 'THOM', 'TOAD',
'TOBA', 'TSAY', 'TSEA', 'TSIT', 'TURN', 'UARL', 'UBIR', 'UBTN', 'UCHR', 'UDEN', 'UEUT', 'UFRA', 'UHAF', 'UISR',
'UKEC', 'UMUS', 'UNAR', 'UNRS', 'UNTH', 'UNUR', 'UOMI', 'UPCE', 'USHU', 'USIK', 'USKE', 'UTRE', 'VICT', 'WILL',
'WORC', 'ZYMO')

-- calculate the area weighted mean of precip contributing to a stream
WITH areas AS
(SELECT
  a.id,
  b.map,
  b.area
FROM bcfishpass.mean_annual_precip a
INNER JOIN bcfishpass.mean_annual_precip b
ON FWA_Upstream(a.wscode_ltree, a.localcode_ltree, b.wscode_ltree, b.localcode_ltree)
WHERE a.watershed_group_code IN :wsg
),
-- WHERE a.watershed_group_code IN ('BULK', 'MORR', 'LSKE', 'KISP', 'KLUM', 'LKEL', 'ZYMO', 'BABL', 'BABR', 'MSKE', 'SUST', 'USKE')
-- WHERE a.watershed_group_code IN ('BULK','LNIC','HORS','ELKR','MORR')


totals AS
(
  SELECT
  a.id,
  SUM(area) as area
  FROM areas a
  GROUP BY a.id
),

weighting AS
(
SELECT
 a.id,
 a.area,
 b.area as area_total,
 a.area / b.area as pct,
 a.map,
 (a.area / b.area) * a.map as weighted_map
FROM areas a
INNER JOIN totals b ON a.id = b.id
),

weighted_mean AS
(
SELECT
  id,
  round(sum(weighted_map)) as map_upstream
FROM weighting
GROUP BY id
)

UPDATE bcfishpass.mean_annual_precip s
SET map_upstream = w.map_upstream
FROM weighted_mean w
WHERE s.id = w.id;

**Comment by smnorris** [comment]
Probably - that isn't something I've tried. You will probably have to fuss with quoting. 
It is redundant but I'd probably put it right in the query to avoid messing with bash variables/strings and just get it working.

**Comment by smnorris** [comment]
Maybe? Can you define the variable right in the sql?!?

**Comment by NewGraphEnvironment** [comment]
doubt it.  I'm not familiar with la lengua.  Where did you define wsg in `load_fwa_assessment_watersheds_lut.sql`?

**Comment by smnorris** [comment]
In the shell script, it queries for all groups:
https://github.com/smnorris/fwapg/blob/main/extras/assessment_watersheds_lookups.sh#L13

**Comment by smnorris** [comment]
I don't particularly like bash scripting but for these few specific tasks it works really well (once I figured out the syntax)

**Comment by NewGraphEnvironment** [comment]
I can't figure it out and I should move on.  I have a feeling that I am going to either run into memory issues or run into other complications trying to insert `map_upstream` into my current `bcfishpass.mean_annual_precip` table as it only has `map` values only about 200 of the 240 watershed groups (any watershed group with 10 or more FISS sites...).

I think the `map_upstream` is a worthy cause so I wonder if, when you get a chance, you might be able to make me a `bcfishpass.mean_annual_precip`  table from scratch for the whole province that I could join to `bcfishpass.channel_width_measured` ? (I have the whole province run for `bcfishpass.channel_width_measured` and have added [extra columns for FISS vs. PSCIS sites](https://github.com/NewGraphEnvironment/bcfishpass/blob/newgraph/01_prep/04_channel_width/sql/channel_width_measured.sql).

**Comment by NewGraphEnvironment** [comment]
I managed to update my `bcfishpass.mean_annual_precip`  (190 or so watershed groups included) with map_upstream using the `-v wsg={1}`  action. Nice one. Thank you. 
 I don't have a lot of hope that it would run but I took a shot at modifying the scripts to run the precip analysis for all the watershed groups from the start. 
 Untested - they are 
[here ]( https://github.com/NewGraphEnvironment/bcfishpass/blob/newgraph/01_prep/04_channel_width/mean_annual_precip_all.sh)
and [here](https://github.com/NewGraphEnvironment/bcfishpass/blob/newgraph/01_prep/04_channel_width/sql/mean_annual_precip_all.sql)

Keen to see how it really should be done.

**Comment by smnorris** [comment]
I think that is pretty much it. I'll test.

**Comment by NewGraphEnvironment** [comment]
Did it run? That time command is neat.

**Comment by smnorris** [comment]
This is all that is needed to do the raster processing provincially, just remove the watershed group selection and it will do all of them: https://github.com/smnorris/bcfishpass/commit/27a4281efea9836ed50f97372abe38fcfce91532

Once that is done processing I'll try changing the call to `sql/mean_annual_precip.sql`

**Comment by NewGraphEnvironment** [comment]
I believe that running the whole province on that raster processing call crashed my system twice in a row.  Interestingly, I was able to get it to finish if I removed a fraction of the watershed groups (I pulled out all watershed groups with less than 10 FISS sites - about 50 or so groups I think).  The processing did jam up my wsl with 150GB of noise though.  I could be mistaken that it was the raster processing that caused the issue but I don' think so because the `bcfishpass.mean_annual_precip`  table did not have map values after the crashes....

Maybe when I get some time later in the month I will try some tests to see if I can break it up.  The way I wrote it probably won't work with multiple parallel calls will it?

**Comment by smnorris** [comment]
Ok. Let me know which command caused the crash.

The existing command already loops through individual watershed groups in parallel, adding this extra query won't really do anything. 
```
psql -t -P border=0,footer=no \
-c "SELECT ''''||watershed_group_code||''''
    FROM whse_basemapping.fwa_watershed_groups_poly
    ORDER BY watershed_group_code" \
  | sed -e '$d' \
  | parallel psql -t -c "SELECT
  ST_AsGeoJSON(t.*)
FROM
  (
    SELECT DISTINCT
      watershed_feature_id,
      watershed_group_code,
      geom
    FROM whse_basemapping.fwa_watersheds_poly
    WHERE watershed_group_code = :wsg
  ) as t" -v wsg={1}| \
  parallel \```

**Comment by NewGraphEnvironment** [comment]
I know for sure that the `map_upstream` insert causes memory overload crashes and I am pretty sure the raster sampling does too.  Are you sure the raster sampling loops through the watershed groups?  Can you maybe point why in the original call below? Is it because it calls distinct `watershed_feature_id` and `watershed_group_code`? Wouldn't that just run through a mulitimillion row list before outputting anything to disk?

```
psql -t -c "SELECT
  ST_AsGeoJSON(t.*)
FROM
  (
    SELECT
      watershed_feature_id,
      watershed_group_code,
      geom
    FROM whse_basemapping.fwa_watersheds_poly
  ) as t" | \
  parallel \
    --pipe \
    "rio -q zonalstats \
      -r mean_annual_precip.tif \
      --prefix 'map_'" | \
  jq '.features[].properties | {watershed_feature_id: .watershed_feature_id, watershed_group_code: .watershed_group_code, map: .map_mean}' | \
  jq --slurp . | \
  in2csv -f json |
  psql -c "\copy bcfishpass.mean_annual_precip_wsd_load FROM STDIN delimiter ',' csv header"
```

**Comment by smnorris** [comment]
It might indeed load all the results into memory before loading to disk, I'm not sure.

**Comment by smnorris** [comment]
Ok, sorry - I wasn't looking closely enough at the query.  Splitting it up like you've done is probably the right approach to make it work if you're memory limited (RAM) but I'm not sure how best to make things flow. I'll just run it and get you the output because adapting will take come reading

**Comment by NewGraphEnvironment** [comment]
ok. Thanks. must be something in the way the RAM is allocated if that is the issue. I have 16GB...

**Comment by NewGraphEnvironment** [comment]
you have 32GB?

**Comment by smnorris** [comment]
64G FTW. Rarely use it all but it is really nice when needed. I can process provincial 10m rasters without issue.

**Comment by NewGraphEnvironment** [comment]
maybe i do need a new setup. Is your unit bought of the shelf?  It is a macbook?

**Comment by smnorris** [comment]
Yep, just a 16" apple laptop. The big RAM add on is expensive but you can't upgrade it later, might as well pay up front.

**Comment by smnorris** [comment]
yup.

**Comment by NewGraphEnvironment** [comment]
this does not seem to run on my machine working off up to date `main` branch with [29a721a commit intact](https://github.com/smnorris/bcfishpass/commit/29a721a1a59d81d3d6d2a34013ec89b5fe5555c4)....  Am I missing something?

```
(bcfpenv) newgraph@MSI:~/scripts/bcfishpass/01_prep/04_channel_width$ ./mean_annual_precip.sh
+ TMP=/home/newgraph/tmp
+ psql -c 'DROP TABLE IF EXISTS bcfishpass.mean_annual_precip_wsd_load'
DROP TABLE
+ psql -c 'CREATE TABLE bcfishpass.mean_annual_precip_wsd_load (watershed_feature_id integer, watershed_group_code text, map numeric)'
CREATE TABLE
++ psql -t -P border=0,footer=no -c 'SELECT watershed_group_code
      FROM whse_basemapping.fwa_watershed_groups_poly
      ORDER BY watershed_group_code'
+ for WSG in $(psql -t -P border=0,footer=no   -c "SELECT watershed_group_code
      FROM whse_basemapping.fwa_watershed_groups_poly
      ORDER BY watershed_group_code")
+ echo 'Processing ADMS'
Processing ADMS
+ psql -X -t -v wsg=ADMS
+ jq '.features[].properties | [.watershed_feature_id, .watershed_group_code, .map_mean]'
+ parallel --pipe 'rio -q zonalstats         -r mean_annual_precip.tif         --prefix '\''map_'\''         2>/dev/null'
+ jq -r --slurp '.[] | @csv'
+ psql -c '\copy bcfishpass.mean_annual_precip_wsd_load FROM STDIN delimiter '\'','\'' csv header'
COPY 0
```

**Comment by smnorris** [comment]
I don't see lines 27-36 in your snippet. Are the watersheds being extracted? 
https://github.com/smnorris/bcfishpass/blob/main/01_prep/04_channel_width/mean_annual_precip.sh#L26

To make sure they are you could test it out on a small set. Below would echo the geojson of the first 10 watersheds in ADMS to your terminal.

```
for WSG in $(psql -t -P border=0,footer=no \
  -c "SELECT watershed_group_code
      FROM whse_basemapping.fwa_watershed_groups_poly
      WHERE watershed_group_code = 'ADMS'
      ORDER BY watershed_group_code")
do
  echo 'Processing '$WSG
  psql -X -t -v wsg="$WSG" <<< "SELECT
    ST_AsGeoJSON(t.*)
  FROM
    (
      SELECT
        watershed_feature_id,
        watershed_group_code,
        geom
      FROM whse_basemapping.fwa_watersheds_poly
      WHERE watershed_group_code = :'wsg'
      LIMIT 10
    ) as t
done "
```
output:
```
 Processing ADMS
 {"type": "Feature", "geometry": {"type":"MultiPolygon","crs":{"type":"name","properties":{"name":"EPSG:3005"}},"coordinates":[[[[1476644.832,762117.684],[1476654.895,762107.948],[1476662.103,762101.195],[1476662.576,762100.616],[1476668.289,762094.907],[1476739.671,762079.787],[1476699.675,762114.428],[1476680.597,762128.811],[1476666.853,762138.144],[1476644.832,762117.684]]]]}, "properties": {"watershed_feature_id": 8500129, "watershed_group_code": "ADMS"}}
```

**Comment by NewGraphEnvironment** [comment]
that works.  I don't get the same error as before but I wonder if my `rasterstats` bailed out of `rasterio` again.  

When I run 

`rio --help` 

I don't see `zonalstats` as an option anymore.  Thinking I might need to rebuild the virt env.  

Do you still see `zonalstats` when you run `rio --help` ?

**Comment by NewGraphEnvironment** [comment]
apparently I need to 

`conda deactivate` out of my `base` virt env then 

`conda activate bcfpenv` for `rasterstats` to plug into `rasterio`.  Seems so odd.  Ok. Its running.  Thanks for the quick reply.

**Comment by smnorris** [comment]
If you re-create the `bcfpenv` conda environment with the latest yml, `bcdata` will now install with the latest and greatest.

**Comment by NewGraphEnvironment** [comment]
👌🏼
